### PR TITLE
seccomp: fix fd garbage collector issue

### PR DESCRIPTION
**Pull request by alban**
Keep a reference on the eBPF link, even if we don't access it.
Otherwise, the garbage collector will free it and call the finalizer set
up by the cilium/ebpf library: [internal/fd.go#L21](https://github.com/cilium/ebpf/blob/4d5607ee1690d067b7f8befdd69b2bbee07e644b/internal/fd.go#L21)
```
runtime.SetFinalizer(fd, (*FD).Close)
```
Symptoms of the bug before this fix: the seccomp gadget works fine but
stops working after a while (when the garbage collector decides to call
the finalizer and close the file descriptor).

### CODE REVIEW

File: pkg/gadgets/seccomp/tracer/tracer.go
Comment (rata) [inline pkg/gadgets/seccomp/tracer/tracer.go:85]: This seems unrelated, but it is harmless :)

**Review by rata — APPROVED** [review]
@alban heh, this GC issue rings a bell :-D

I'm of course not familiar with the code to know the lifetime of the Tracer t, but on a cursory look LGTM. But please do wait on reviews by others :)

And just curious: do you know if upstream really wants to have a finalizer on the object? From the commit msg it seems like probably they do, but I've seen similar code in runc that didn't really want a finalizer. So, wondering if upstream wants a finalizer taking into account this caveats. It will be nice to contribute upstream :)

**Comment by alban** [comment]
@rata This seems to be done on purpose because the FD struct has a `Forget()` method to remove the finalizer if needed See also commit messages from:

https://github.com/cilium/ebpf/commit/b9ee7bc3f04d5b2721892c1e2786f31d9f7a9b0f
https://github.com/cilium/ebpf/commit/f4398602ca2a37b99a1f29df9a7e8adcc57be680

I think I forgot to call `t.progLink.Close()` explicitly in this patch, I'll check.

File: pkg/gadgets/seccomp/tracer/tracer.go
Comment (alban) [reply pkg/gadgets/seccomp/tracer/tracer.go:85]: This error case happened when `progLink` was closed by the garbage collector, so the error message will help to detect this if this bug happens again.

**Comment by alban** [comment]
I rebased and added the missing `t.progLink.Close()`.

File: pkg/gadgets/seccomp/tracer/tracer.go
Comment (rata) [reply pkg/gadgets/seccomp/tracer/tracer.go:85]: Ohh, cool!

**Review by mauriciovasquezbernal — APPROVED** [review]
Nice finding. What about adding a comment in L63 saying we keep this reference to avoid the GC messing up?

**Comment by rata** [comment]
@mauriciovasquezbernal +1. Or maybe in the struct field?

**Review by marga-kinvolk — APPROVED** [review]
LGTM

File: pkg/gadgets/seccomp/tracer/tracer.go
Comment (rata) [inline pkg/gadgets/seccomp/tracer/tracer.go:39]:
```suggestion
	// A reference is kept so it can be closed it explicitly, otherwise
```

You may want to add a linter that checks this (runc has one) in the future :)

**Review by rata — APPROVED** [review]
Small nit, but still LGTM :)

**Pull request merged**
